homeassistant/components/mychevy.py

```diff
@@ -0,0 +1,343 @@
+        self._car = None
+        self.add_devices = None
+        self.sensors = []
+        # This is a status sensor for the connection itself
+        self.status = MyChevyStatus(self)
+
+    @property
+    def car(self):
+        """An instance of mychevy.mychevy.EVCar."""
+        return self._car
+
+    @car.setter
+    def car(self, data):
+        """Update the EVCar.
+
+        Also update all linked sensors in hass and signal the platform
+        there are updates.
+
+        """
+        self._car = data
+        for sensor in self.sensors:
+            sensor.car = data
+            sensor.schedule_update_ha_state()
+
+    def register_add_devices(self, add_devices):
+        """Register add_devices method.
+
+        This creates the initial status sensor, and makes it possible
+        for the hub to spin up devices as it has appropriate
+        information for them.
+
+        """
+        self.add_devices = add_devices
+        add_devices([self.status], True)
+
+    def _create_sensors(self):
+        """Create the sensors for the car.
+
+        This is not done during initialization because the sensors
+        have no value at initialization time.
+
+        """
+        self.sensors = [
+            EVRange(self),
+            EVMileage(self),
+            EVCharge(self),
+            EVPlugged(self),
+            EVCharging(self),
+            EVChargeMode(self)
+        ]
+        add = self.add_devices
+        add(self.sensors, True)
+
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
+    def update(self):
+        """Update sensors from mychevy website.
+
+        This is a synchronous polling call.
+        """
+        self.car = self._client.data()
+        self.status.success()
+        if self.add_devices and not self.sensors:
+            self._create_sensors()
+
+    def run(self):
+        """Thread run loop."""
+        # We add the status device first outside of the loop
+
+        # And then busy wait on threads
+        while True:
+            try:
+                _LOGGER.info("Starting mychevy loop")
+                self.update()
+                time.sleep(MIN_TIME_BETWEEN_UPDATES.seconds)
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception(
+                    "Error updating mychevy data. "
+                    "This probably means the OnStar link is down again")
+                self.status.error()
+                time.sleep(ERROR_SLEEP_TIME)
+
+
+class EVSensor(Entity):
+    """Base EVSensor class.
+
+    The only real difference between sensors is which units and what
+    attribute from the car object they are returning. All logic can be
+    built with just setting subclass attributes.
+
+    """
+
+    _icon = None
+    _unit = "miles"
+    _name = "EV Sensor"
+    _attr = None
+
+    def __init__(self, connection):
+        """Initialize sensor with car connection."""
+        self._conn = connection
+        self.hass = connection.hass
```

Comment:
Home assistant will set `hass` on the entity when it has been added. So remove this.

Reply:
ack